src/clients/BalanceAllocator.ts

```diff
@@ -82,15 +93,15 @@ export class BalanceAllocator {
 
   async getBalance(chainId: number, token: string, holder: string) {
     if (!this.balances?.[chainId]?.[token]?.[holder]) {
-      const balance =
-        token === ZERO_ADDRESS
-          ? await this.providers[chainId].getBalance(holder)
-          : await ERC20.connect(token, this.providers[chainId]).balanceOf(holder);
-
-      // Note: cannot use assign because it breaks the BigNumber object.
-      if (!this.balances[chainId]) this.balances[chainId] = {};
-      if (!this.balances[chainId][token]) this.balances[chainId][token] = {};
-      this.balances[chainId][token][holder] = balance;
+      const balance = await this._queryBalance(chainId, token, holder);
+      // To avoid inconsitencies, we recheck the balances value after the query.
+      // If it exists, skip the assignment so the value doesn't change after being set.
+      if (!this.balances?.[chainId]?.[token]?.[holder]) {
+        // Note: cannot use assign because it breaks the BigNumber object.
+        if (!this.balances[chainId]) this.balances[chainId] = {};
```

Comment:
Minor: wdyt about using the ??= operator here? It seems like a more concise way of doing this and it avoids the need to repeat the lookup path into the object, so it _might_ be less prone to copy/paste errors creeping in with future changes.

nb. I should note that the `if (!x[a][b]) x[a][b] = {};` pattern is pre-existing in the code and it appears in several other places that are not touched by this change, so if it makes sense to apply this change then maybe it could be done in a follow-up commit to avoid polluting this one?

```suggestion
        this.balances[chainId] ??= {};
```

Reply:
I think it's going to be tough to root out all the places where we have `if(!x) x = y` throughout the codebase (and it's somewhat risky), but I think using this simpler syntax is great and we should do it when modifying/adding code for sure!!